execute/executor.go

```diff
@@ -287,17 +300,49 @@ func (v *createExecutionNodeVisitor) Visit(node plan.Node) error {
 				}
 			}
 
-			if (plan.HasSideEffect(spec) || isParallelMerge) && len(node.Successors()) == 0 {
-				name := string(node.ID())
-				r := newResult(name)
-				v.es.results[name] = r
-				v.nodes[skipYields(node)][i].AddTransformation(r)
+			// XXX: results should be generated for terminal nodes, regardless
+			// of whether or not it is also a leaf. This is why there's an
+			// identical block at the end of each loop in each case for the
+			// "if leaf" check.
+			if len(node.Successors()) == 0 {
+				resultName := getResultName(node, spec, isParallelMerge)
+				if err := v.generateResult(resultName, node, i); err != nil {
+					return err
+				}
 			}
 		}
 	}
 	return nil
 }
 
+// generateResult will attach a result to the query for the specified node.
+func (v *createExecutionNodeVisitor) generateResult(resultName string, node plan.Node, idx int) error {
+	// if the result name is already present in the result set, that's an error.
+	if _, ok := v.es.results[resultName]; ok {
+		// XXX: we produce an error like this in the planner for duplicate yield
+		// names, but since we're generating results that aren't necessarily from
+		// yields, we need a similar check here.
+		return errors.Newf(codes.Invalid, "tried to produce more than one result with the name %q", resultName)
+	}
+	r := newResult(resultName)
+	v.es.results[resultName] = r
+	v.nodes[skipYields(node)][idx].AddTransformation(r)
+	return nil
+}
+
+// getResultName will offer a "best guess" name for a given node's result.
+//
+// For nodes that have side-effects or happen to be a Parallel Merge, the result
+// will be based on the node ID. For other cases, the default yield name of
+// `_result` is used.
+func getResultName(node plan.Node, spec plan.ProcedureSpec, isParallelMerge bool) string {
+	name := plan.DefaultYieldName
+	if plan.HasSideEffect(spec) || isParallelMerge {
+		name = string(node.ID())
+	}
+	return name
+}
```

Comment:
Is this a departure from what we did before? It seems like this logic is carefully constructed such that it's not.

Reply:
Hard to answer as yes or no. This branching behavior matches the _outcomes_ we saw given the logic up in the planner previously _combined with_ the logic we had in the executor to start.

The old implicit yields were all named using the default yield name. The side-effects and yields that were _not implicit_ used the node id and the name associated with the yield respectively.

When we combine all this with the 3 calls to `generateResult` (yields using their own names, the rest using this logic) we should match the original outcome.

Reply:
> It seems like this logic is carefully constructed such that it's not.

@wolffcm follow-up question for you on this: is there a different scheme that's been discussed that we'd like to implement?

Reply:
Since we handle literal yields separately (and always use the name they carry with them) we should be able to move the `string(node.ID())` calls to the  callsite and remove this function. This would make me happy since it looks odd out of the historical context.

The net change for this would be: _"what were **formerly implicit yields** will no longer use `_result` for their name, instead using the ID of the node that produced them."_

Reply:
My reading of the code is that previously we would use `_result` only for nodes with no side effects.

I think the idea here is that if you just had `from |> range |> filter` but forgot to add the yield, we would generate one for you, and give it the default name. I am rethinking my suggestion since this seems like reasonable behavior to me.

I'm looking at the comment that added result generation in the execution graph, here:
https://github.com/influxdata/flux/commit/983a88e597de5ee5666fa23159d83e501ceef4e3

It seems like there's a technical reason that we need to add the result---the would terminate early otherwise. It seems like maybe this was a quick fix to a deep problem. Maybe for that case should add an implicit `debug.sink()`?!? I am only half-joking.

I guess for now I'd be more comfortable leaving the current behavior as-is.

Reply:
Sorry for the flip-flopping on that... what you have here is still a great simplification! Could I trouble you to add what I wrote as a comment somewhere in `executor.go`?

> If user entered `from |> range |> filter` but forgot to add the yield, we generate one for them, and give it the default name.

> We also generate a result for terminal nodes that have side effects, since all
queries must have an associated result transformation, otherwise
the query context will be cancelled immediately and execution
will be cancelled before any work for the query has been done.
TODO: understand is this could be addressed by another means.

Reply:
I just want to be sure that we capture the rationale for the naming.

Reply:
Captured with 4642d255ced6870e6d028bf222b06d1debf3ff5f